drivers/volume/portworx/portworx.go

```diff
@@ -642,22 +641,40 @@ func (d *portworx) GetNodeForVolume(vol *torpedovolume.Volume) (*node.Node, erro
 		}
 	}
 
-	pxVol := v.(*api.Volume)
-	for _, n := range node.GetStorageDriverNodes() {
-		if n.VolDriverNodeID == pxVol.AttachedOn {
-			return &n, nil
+	r := func() (interface{}, bool, error) {
+		pxVol := v.(*api.Volume)
+		for _, n := range node.GetStorageDriverNodes() {
+			if n.VolDriverNodeID == pxVol.AttachedOn {
+				return &n, false, nil
+			}
+		}
+
+		// Snapshots may not be attached to a node
+		if pxVol.Source.Parent != "" {
+			logrus.Debugf("Volume with ID: %s is a snapshot", pxVol.Id)
+			return nil, false, nil
+		}
+
+		return nil, true, &ErrFailedToInspectVolume{
+			ID:    name,
+			Cause: "Volume is not attached on any node",
 		}
 	}
 
-	// Snapshots may not be attached to a node
-	if pxVol.Source.Parent != "" {
-		return nil, nil
+	n, err := task.DoRetryWithTimeout(r, defaultTimeout, defaultRetryInterval)
+	if err != nil {
+		return nil, &ErrFailedToInspectVolume{
+			ID:    name,
+			Cause: err.Error(),
+		}
 	}
 
-	return nil, &ErrFailedToInspectVolume{
-		ID:    name,
-		Cause: "Volume is not attached on any node",
+	if n != nil {
+		node := n.(*node.Node)
+		return node, nil
 	}
+
+	return nil, nil
```

Comment:
the default return should return an error just like it was doing before. An API should not be returning a nil err and also a nil return object. The user of the function wouldn't know what to do with it.

Reply:
discussed offline with nikolay. The current implementation is correct.